### PR TITLE
done BUZ-131, add tests, add experimental Proxy validation

### PR DESCRIPTION
**Pull request by tttmaximttt**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -39,3 +39,7 @@ jspm_packages
 # Test Environment Variable File
 test.env
 .env
+.idea/
+.idea/buzzi-sdk-node.iml
+.idea/modules.xml
+.idea/workspace.xml
```
Comment (DerekSeverson) [inline .gitignore]: @tttmaximttt just gitignore `.idea` which will ignore the entire folder and everything inside.

File: lib/service.js [hunk 2]
```diff
@@ -12,6 +17,13 @@ class Service {
     this.host = host || process.env.BUZZI_HOST || 'https://core.buzzi.io';
     this.id = id || process.env.BUZZI_API_ID;
     this.secret = secret || process.env.BUZZI_API_SECRET;
+    const validator = new ServiceValidator();
+    const proxy = new Proxy(this, {
+      get(target, key) {
+        return validator[key] ? validator[key](target, key) : target[key];
+      }
+    });
+    return proxy;
```
Comment (DerekSeverson) [inline lib/service.js]: @tttmaximttt - I would remove the Proxy validator functionality for now and keep it simple.  Just put the validation and parameter processing in the same "uploadFiles" function.

File: lib/service.js [hunk 3]
```diff
@@ -81,6 +93,26 @@ class Service {
...
+  /**
+   *
+   * @param {string|string[]} files
+   * @returns {Promise}
+   */
+  uploadFiles(files) {
+    const formData = { attachments: [] };
+
+    for(const file of files) {
+      const stream = fs.createReadStream(file);
```
Comment (DerekSeverson) [inline lib/service.js]: @tttmaximttt - we would rather accept an array of "Buffers" and "ReadStreams".   We should not be creating the readstream from a file path ourselves.  The Developer (SDK user) should create his/her own buffers/readstreams from files (or anything else for that matter), and then pass that to us.  Add validation to ensure we only are excepting an arrays that have Buffers and ReadStreams.  
class__aws_s3_ _aws_sdk_for_javascript

File: lib/validators/service.validator.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+const {
```
Comment (DerekSeverson) [inline lib/validators/service.validator.js]: @tttmaximttt - delete.  Just add validation to the method itself.

**Review by DerekSeverson — CHANGES_REQUESTED** [review]
See inline comments.

File: .gitignore (hunk 1 above)
Comment (tttmaximttt) [reply .gitignore]: fix

File: lib/service.js (hunk 2 above)
Comment (tttmaximttt) [reply lib/service.js]: fix

File: lib/service.js (hunk 3 above)
Comment (tttmaximttt) [reply lib/service.js]: fix

File: lib/validators/service.validator.js (hunk 4 above)
Comment (tttmaximttt) [reply lib/validators/service.validator.js]: fix

File: test/service.js [hunk 5]
```diff
@@ -51,33 +51,74 @@ describe('Service', function () {
...
       const stub = sinon.stub(service, 'request');
       const fixturePath1 = path.join(__dirname, 'fixtures/files/hello.txt');
       const fixturePath2 = path.join(__dirname, 'fixtures/files/image.jpg');
-      service.uploadFiles([fixturePath1, fixturePath2]);
+      const formData = {
+        attachments: [
+          fs.createReadStream(fixturePath1),
+          fs.createReadStream(fixturePath2),
+        ],
+        bufFile:  new Buffer('asfasfas'),
```
Comment (DerekSeverson) [inline test/service.js]: Using the Buffer constructor is frowned upon and is deprecated:  https://nodejs.org/api/buffer.html#buffer_new_buffer_string_encoding
Use `Buffer.from(string)` method instead.
https://nodejs.org/api/buffer.html#buffer_class_method_buffer_from_string_encoding

File: lib/util.js [hunk 6]
```diff
@@ -1,5 +1,5 @@
 'use strict';
-
+const { Readable,  Duplex, Transform } = require('stream');
```
Comment (DerekSeverson) [inline lib/util.js]: eslint should have shown errors for "unused variables".

File: lib/service.js [hunk 7]
```diff
@@ -93,26 +88,31 @@ class Service {
     });
   }
 
+  static _checFormData(formData) {
```
Comment (DerekSeverson) [inline lib/service.js]: Extra function not needed, just add validation into `upload` method itself. 
Keep the interface of the `Service` class as clean/minimal as possible without extra methods - even if its an underscore/private method.

File: lib/service.js [hunk 8]
```diff
@@ -93,26 +88,31 @@ class Service {
...
+    if(isEmpty(formData)) {
+      throw new Error('Can\'t be empty');
+    }
+
+    if (!isObject(formData)) {
+      throw new Error('Should be an object');
+    }
+  }
```
Comment (DerekSeverson) [inline lib/service.js]: Iterate through the formData's values and validation each of them to make sure they have proper values as well.

File: lib/service.js [hunk 9]
```diff
@@ -93,26 +88,31 @@ class Service {
...
   /**
    *
-   * @param {string|string[]} files
+   * @param {object} formData
+   * [Request formData]{@link https://github.com/request/request#forms}
```
Comment (DerekSeverson) [inline lib/service.js]: cool!

File: lib/service.js [hunk 10]
```diff
@@ -1,11 +1,13 @@
 'use strict';
 
-const ServiceValidator = require('./validators/service.validator');
+const {
+  isEmpty,
+  isObject,
+} = require('lodash');
```
Comment (DerekSeverson) [inline lib/service.js]: wait ... we don't have `lodash` as a dependency.  I don't believe these can be used here.  For now, I think we're wanting to get by without using lodash ... trying to keep dependencies to a minimum.  We may need to later, but not for now.

**Review by DerekSeverson — CHANGES_REQUESTED** [review]
a few more changes needed.

File: lib/service.js (hunk 8 above)
Comment (tttmaximttt) [reply lib/service.js]: But what a proper values for formData, in formData we can have any types o fields, what exactly I should check?

Comment (tttmaximttt) [reply lib/service.js]: And a part of validation realized in request lib.

Comment (DerekSeverson) [reply lib/service.js]: Sounds good.  I agree.

**Review by DerekSeverson — APPROVED** [review]
Looks good.

**Pull request merged**